python/datasources/uhc.py

```diff
@@ -0,0 +1,122 @@
+    'Hawaiian/Pacific Islander': Race.NHPI_NH.value,
+    'Other Race': Race.OTHER_STANDARD_NH.value,
+    'White': Race.WHITE_NH.value,
+    'Multiracial': Race.MULTI_NH.value,
+    'All': Race.ALL.value,
+}
+
+BASE_UHC_URL = "https://www.americashealthrankings.org/api/v1/downloads/210"
+
+UHC_DETERMINANTS_OF_HEALTH = {
+    "Chronic Obstructive Pulmonary Disease": std_col.COPD_PCT,
+    "Diabetes": std_col.DIABETES_PCT,
+}
+
+
+class UHCData(DataSource):
+
+    @staticmethod
+    def get_id():
+        return 'UHC_DATA'
+
+    @staticmethod
+    def get_table_name():
+        return 'uhc_data'
+
+    def upload_to_gcs(self, _, **attrs):
+        raise NotImplementedError(
+            'upload_to_gcs should not be called for UHCData')
+
+    def write_to_bq(self, dataset, gcs_bucket, **attrs):
+        df = gcs_to_bq_util.load_csv_as_dataframe_from_web(
+                BASE_UHC_URL, dtype={std_col.STATE_FIPS_COL: str}
+            )
+
+        for b in [std_col.RACE_OR_HISPANIC_COL, std_col.AGE_COL, std_col.SEX_COL]:
+            breakdown_df = self.generate_breakdown(b, df)
+
+            column_types = {c: 'STRING' for c in breakdown_df.columns}
+            for col in [std_col.COPD_PCT, std_col.DIABETES_PCT]:
+                column_types[col] = 'FLOAT'
+
+            if std_col.RACE_INCLUDES_HISPANIC_COL in breakdown_df.columns:
+                column_types[std_col.RACE_INCLUDES_HISPANIC_COL] = 'BOOL'
+
+            breakdown_df.to_csv("test_%s.csv" % b)
+            gcs_to_bq_util.add_dataframe_to_bq(
+                breakdown_df, dataset, b, column_types=column_types)
+
+    def generate_breakdown(self, breakdown, df):
+        breakdown_map = {
+            "race_and_ethnicity": UHC_RACE_GROUPS,
+            "age": UHC_AGE_GROUPS,
+            "sex": UHC_SEX_GROUPS,
+        }
+
+        output = pd.DataFrame()
+        states = df['State Name'].drop_duplicates().to_list()
+
+        for state in states:
+            for breakdown_value in breakdown_map[breakdown]:
+                output_row = {}
+                output_row['state_name'] = state
+
+                if breakdown == std_col.RACE_OR_HISPANIC_COL:
+                    output_row[std_col.RACE_CATEGORY_ID_COL] = UHC_RACE_GROUPS_TO_STANDARD[breakdown_value]
+                else:
+                    output_row[breakdown] = breakdown_value
+
+                output_row['state_fips'] = constants.STATE_NAMES_TO_FIPS[state]
+
+                for determinent in UHC_DETERMINANTS_OF_HEALTH:
+                    if breakdown_value == 'All':
+                        output_row[UHC_DETERMINANTS_OF_HEALTH[determinent]] = df.loc[
+                            (df['State Name'] == state) &
+                            (df['Measure Name'] == determinent)]['Value'].values[0]
+
+                    else:
+                        row = df.loc[
+                            (df['State Name'] == state) &
+                            (df['Measure Name'].str.contains(determinent)) &
```

Comment:
I'm not a huge fan of doing str.contains to find the right entry in the df, but the raw data is in kind of a weird format. What do you think about building the correct phrase, e.g. "Diabetes - White", and then looking it up directly?

Reply:
We would have to add another if condition to special case for age in that case. TBH I would rather avoid that and leave it like this.

Reply:
SG